implement a2dp remain todo/interfaces

MarkWangChinese opened the pull request:
Implement the remain A2DP todo/interfaces as follow:

Implement the bt_a2dp_stream_reconfig: Modify reconfig_req callback to
pass codec_cfg to application. Remove reconfig_rsp callback,
config_rsp is used. Remove reconfigured callback, configured callback is used.

Implement the bt_a2dp_stream_release: The release_req,
release_rsp and released callbacks are already defined,
implement them in this change.

Implement the bt_a2dp_stream_suspend: The suspend_req,
suspend_rsp and suspended callbacks are already defined,
implement them in this change.

Add and implement the bt_a2dp_stream_abort: Add abort_req,
abort_rsp and aborted callbacks.

Implement bt_a2dp_disconnect.

Simplify the codes: (1) use the same bt_avdtp_ctrl_params to process open,
start, suspend, close and abort operations. (2) use same a2dp_ctrl_ind to
process open, start, suspend, close and abort operations. (3) use the same
bt_a2dp_ctrl_cb to process open, start, suspend, close and abort operations.
(4) use the same bt_avdtp_ctrl to process open, start, suspend, close and
abort operations. (5) move the status to common struct bt_avdtp_req.

Add shell test.

jhedberg commented on subsys/bluetooth/host/classic/a2dp.c:
```diff
@@ -199,6 +196,10 @@ static int a2dp_set_config_ind(struct bt_avdtp *session, struct bt_avdtp_sep *se
 	__ASSERT(sep, "Invalid sep");
 
 	ep = CONTAINER_OF(sep, struct bt_a2dp_ep, sep);
+	if (ep == NULL) {
```
NULL checks for `CONTAINER_OF()` are almost always wrong. It's not a function, rather just pointer arithmetic. The only way you'd get NULL is if you give it a pointer which isn't NULL but has exactly the value `NULL + offsetof(struct bt_a2dp_ep, sep)`. Checking for something like that seems pretty odd to me.

MarkWangChinese replied:
Removed the cases and add some more BT_ASSERT for parameter `sep`.

jhedberg commented on subsys/bluetooth/host/classic/a2dp.c:
```diff
@@ -196,7 +194,7 @@ static int a2dp_set_config_ind(struct bt_avdtp *session, struct bt_avdtp_sep *se
 	*errcode = 0;
 
-	__ASSERT(sep, "Invalid sep");
+	BT_ASSERT_MSG(sep, "Invalid sep");
```
What's the reason for changing to `BT_ASSERT()`? My preference would be that we completely remove that API from the Bluetooth subsystem and only use the Zephyr-wide assert API.

MarkWangChinese replied:
Updated all assert

jhedberg commented on subsys/bluetooth/host/classic/a2dp.c:
```diff
@@ -285,166 +321,104 @@ static void bt_a2dp_media_data_callback(
+	struct bt_a2dp_ep *ep = CONTAINER_OF(sep, struct bt_a2dp_ep, sep);
+
+	BT_ASSERT_MSG(sep, "Invalid sep");
+	return a2dp_ctrl_ind(session, sep, errcode,
+		a2dp_cb != NULL ? a2dp_cb->establish_req : NULL,
+		(ep->stream != NULL && ep->stream->ops != NULL) ?
+		ep->stream->ops->established : NULL,
+		false);
```
Sorry, but this is just unreadable. Could you introduce some helper variables which you prepare before the call to `a2dp_ctrl_ind()`, so that the actual call itself doesn't contain any conditionals. The resulting compiled code will most like be exactly the same since the compiler will optimize this either way. Also, please make sure you're using correct alignment of N tabs + 0-7 spaces, so that the split lines start right after the ( on the first line.

MarkWangChinese replied:
Updated all the cases.

jhedberg commented on subsys/bluetooth/host/classic/a2dp.c:
```diff
@@ -285,166 +321,104 @@ static void bt_a2dp_media_data_callback(
+	struct bt_a2dp_ep *ep = CONTAINER_OF(sep, struct bt_a2dp_ep, sep);
+
+	BT_ASSERT_MSG(sep, "Invalid sep");
+	return a2dp_ctrl_ind(session, sep, errcode,
+		a2dp_cb != NULL ? a2dp_cb->start_req : NULL,
+		(ep->stream != NULL && ep->stream->ops != NULL) ?
+		ep->stream->ops->started : NULL,
+		false);
```
Same here.

jhedberg commented on subsys/bluetooth/host/classic/a2dp.c:
```diff
@@ -285,166 +321,104 @@ static void bt_a2dp_media_data_callback(
+	struct bt_a2dp_ep *ep = CONTAINER_OF(sep, struct bt_a2dp_ep, sep);
+
+	BT_ASSERT_MSG(sep, "Invalid sep");
+	return a2dp_ctrl_ind(session, sep, errcode,
+		a2dp_cb != NULL ? a2dp_cb->suspend_req : NULL,
+		(ep->stream != NULL && ep->stream->ops != NULL) ?
+		ep->stream->ops->suspended : NULL,
+		false);
```
And here.

jhedberg commented on subsys/bluetooth/host/classic/a2dp.c:
```diff
@@ -285,166 +321,104 @@ static void bt_a2dp_media_data_callback(
+	struct bt_a2dp_ep *ep = CONTAINER_OF(sep, struct bt_a2dp_ep, sep);
+
+	BT_ASSERT_MSG(sep, "Invalid sep");
+	return a2dp_ctrl_ind(session, sep, errcode,
+		a2dp_cb != NULL ? a2dp_cb->release_req : NULL,
+		(ep->stream != NULL && ep->stream->ops != NULL) ?
+		ep->stream->ops->released : NULL,
+		true);
```
And here.

jhedberg commented on subsys/bluetooth/host/classic/a2dp.c:
```diff
@@ -285,166 +321,104 @@ static void bt_a2dp_media_data_callback(
+	struct bt_a2dp_ep *ep = CONTAINER_OF(sep, struct bt_a2dp_ep, sep);
+
+	BT_ASSERT_MSG(sep, "Invalid sep");
+	return a2dp_ctrl_ind(session, sep, errcode,
+		a2dp_cb != NULL ? a2dp_cb->abort_req : NULL,
+		(ep->stream != NULL && ep->stream->ops != NULL) ?
+		ep->stream->ops->aborted : NULL,
+		true);
```
And here.

jhedberg commented on subsys/bluetooth/host/classic/a2dp.c:
```diff
@@ -703,40 +673,93 @@ int bt_a2dp_stream_config(struct bt_a2dp *a2dp, struct bt_a2dp_stream *stream,
-	return -ENOTSUP;
+	struct bt_a2dp_ep *ep = CONTAINER_OF(CTRL_REQ(req)->sep, struct bt_a2dp_ep, sep);
+
+	return bt_a2dp_ctrl_cb(req, a2dp_cb != NULL ? a2dp_cb->establish_rsp : NULL,
+		(ep->stream != NULL && ep->stream->ops != NULL) ?
+		ep->stream->ops->established : NULL,
+		false);
```
And here.

jhedberg commented on subsys/bluetooth/host/classic/a2dp.c:
```diff
@@ -703,40 +673,93 @@ int bt_a2dp_stream_config(struct bt_a2dp *a2dp, struct bt_a2dp_stream *stream,
+{
+	struct bt_a2dp_ep *ep = CONTAINER_OF(CTRL_REQ(req)->sep, struct bt_a2dp_ep, sep);
+
+	return bt_a2dp_ctrl_cb(req, a2dp_cb != NULL ? a2dp_cb->start_rsp : NULL,
+		(ep->stream != NULL && ep->stream->ops != NULL) ?
+		ep->stream->ops->started : NULL,
+		false);
+}
```
And here

jhedberg commented on subsys/bluetooth/host/classic/a2dp.c:
```diff
@@ -703,40 +673,93 @@ int bt_a2dp_stream_config(struct bt_a2dp *a2dp, struct bt_a2dp_stream *stream,
+{
+	struct bt_a2dp_ep *ep = CONTAINER_OF(CTRL_REQ(req)->sep, struct bt_a2dp_ep, sep);
+
+	return bt_a2dp_ctrl_cb(req, a2dp_cb != NULL ? a2dp_cb->suspend_rsp : NULL,
+		(ep->stream != NULL && ep->stream->ops != NULL) ?
+		ep->stream->ops->suspended : NULL,
+		false);
```
And here

jhedberg commented on subsys/bluetooth/host/classic/a2dp.c:
```diff
@@ -703,40 +673,93 @@ int bt_a2dp_stream_config(struct bt_a2dp *a2dp, struct bt_a2dp_stream *stream,
+{
+	struct bt_a2dp_ep *ep = CONTAINER_OF(CTRL_REQ(req)->sep, struct bt_a2dp_ep, sep);
+
+	return bt_a2dp_ctrl_cb(req, a2dp_cb != NULL ? a2dp_cb->release_rsp : NULL,
+		(ep->stream != NULL && ep->stream->ops != NULL) ?
+		ep->stream->ops->released : NULL,
+		true);
```
And here

jhedberg commented on subsys/bluetooth/host/classic/a2dp.c:
```diff
@@ -703,40 +673,93 @@ int bt_a2dp_stream_config(struct bt_a2dp *a2dp, struct bt_a2dp_stream *stream,
+{
+	struct bt_a2dp_ep *ep = CONTAINER_OF(CTRL_REQ(req)->sep, struct bt_a2dp_ep, sep);
+
+	return bt_a2dp_ctrl_cb(req, a2dp_cb != NULL ? a2dp_cb->abort_rsp : NULL,
+		(ep->stream != NULL && ep->stream->ops != NULL) ?
+		ep->stream->ops->aborted : NULL,
+		true);
```
And here

jhedberg commented on subsys/bluetooth/host/classic/a2dp.c:
```diff
@@ -745,23 +768,85 @@ int bt_a2dp_stream_start(struct bt_a2dp_stream *stream)
+		return -EINVAL;
+	}
+
+	bt_a2dp_stream_config_set_param(stream->a2dp, config, bt_a2dp_set_config_cb,
+			stream->remote_ep != NULL ? stream->remote_ep->sep.sep_info.id :
+			stream->remote_ep_id,
+			stream->local_ep->sep.sep_info.id,
+			stream->local_ep->codec_type, &stream->local_ep->sep);
```
And here

jhedberg commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -58,16 +57,26 @@ K_MUTEX_DEFINE(avdtp_mutex);
-	AVDTP_STREAMING,
-	AVDTP_CLOSING,
-	AVDTP_ABORTING,
+	AVDTP_OPENING = 0x04U,
+	AVDTP_OPEN = 0x08U,
+	AVDTP_STREAMING = 0x10U,
+	AVDTP_CLOSING = 0x20U,
+	AVDTP_ABORTING = 0x40U,
 };
 
+static void bt_avdtp_set_state(struct bt_avdtp_sep *sep, uint8_t state)
+{
+	sep->state = state;
+	if (state != AVDTP_IDLE) {
+		sep->sep_info.inuse = 1u;
+	} else {
+		sep->sep_info.inuse = 0u;
+	}
```
Why are you using capital `U` in some places but lower-case `u` in others?

MarkWangChinese replied:
Change all to `U`

jhedberg commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -729,15 +845,59 @@ static void avdtp_abort_handler(struct bt_avdtp *session,
+		} else if (msg_type == BT_AVDTP_GEN_REJECT) {
+			req->status = BT_AVDTP_NOT_SUPPORTED_COMMAND;
+		}
+		AVDTP_LOCK();
```
You should really fix this global lock stuff. First of all, a semaphore would be better than a mutex, and second of all it should be per-context rather than a global variable.

MarkWangChinese replied:
change to semaphore and move it to the avdtp instance struct.

jhedberg commented on include/zephyr/bluetooth/classic/a2dp.h:
```diff
@@ -30,34 +30,33 @@ extern "C" {
 /** MPEG2,4 IE length */
 #define BT_A2DP_MPEG_2_4_IE_LENGTH (6u)
 /** The max IE (Codec Info Element) length */
-#define A2DP_MAX_IE_LENGTH (8U)
+#define A2DP_MAX_IE_LENGTH         (8U)
```
There's still inconsistency here. Just look at the defines above this. They all use lower-case `u`. Furthermore, there's also a namespace violation: this is a public Bluetooth header file, so all defines and symbols must use a `BT_*` or `bt_*` prefix.

MarkWangChinese replied:
fixed.

jhedberg replied:
> fixed.

No, there's still `A2DP_MAX_IE_LENGTH` which doesn't have a `BT_` prefix.

MarkWangChinese commented on subsys/bluetooth/host/classic/avdtp_internal.h:
```diff
@@ -93,19 +93,19 @@ struct bt_avdtp_sep_info;
-	uint8_t rfa1:3;
-	uint8_t tsep:1;
-	uint8_t media_type:4;
+	uint8_t rfa0: 1;
+	uint8_t inuse: 1;
+	uint8_t id: 6;
+	uint8_t rfa1: 3;
+	uint8_t tsep: 1;
+	uint8_t media_type: 4;
```
clang-format do it. The CI suggest me to do the clang-format.

jhedberg replied:
Ok. That's interesting. Fair enough then, but make sure that style changes to existing code stays in separate commits. Also explain the reasoning in the commit message of such commits.

MarkWangChinese replied:
OK, I will do as this way next time.

jhedberg commented on subsys/bluetooth/host/classic/a2dp.c:
```diff
@@ -99,27 +99,41 @@ static struct bt_a2dp *get_new_connection(struct bt_conn *conn)
 
+	A2DP_LOCK();
 	/* Find a space */
 	for (i = 0; i < CONFIG_BT_MAX_CONN; i++) {
```
You're making this much more complicated than it needs to be. Since you have an array of contexts with `CONFIG_BT_MAX_CONN` elements, you can simply create a 1:1 mapping from `bt_conn` to that array with the help of `bt_conn_index()`: https://docs.zephyrproject.org/latest/doxygen/html/group__bt__conn.html#gad4aed76b80cc815f748ad0e84ae3d87c

I.e. something like:
```c
static struct bt_a2dp *get_new_connection(struct bt_conn_conn)
{
	return &connection[bt_conn_index(conn)];
}
```

I don't think you'll need the `A2DP_LOCK()` stuff after that either.

MarkWangChinese replied:
Theoretically more than one A2DP L2CAP connections can be created based on one ACL connection. Then they are not 1:1 mapping. Usually there is only one A2DP L2CAP connection for one ACL connection. If we need to support it, then the 1:1 mapping cannot be used.

jhedberg replied:
Yes, I know, but can't you just make the general decision (at least for now) that you only support a single A2DP connection per ACL? It seems like you're already doing it, since the code is simply returning the same context if it already exists:
```
	if (connection[i].used == 1U && connection[i].session.br_chan.chan.conn == conn) {
		LOG_DBG("Conn already exists");
		A2DP_UNLOCK();
		return &connection[i];
	}
```

MarkWangChinese replied:
I understand you, but I don't think it is good to limit the A2DP function in order to let the codes simple. I can fix the above codes that you mention. Is it OK?

jhedberg replied:
I'm struggling to understand your point

1. What I'm proposing is not adding any new limits over what you already have in the PR
2. I don't see how you could have multiple "sessions" (i.e. multiple signaling channels) over a single ACL, since the AVDTP spec doesn't define anything like that (AFAIK). It defines the initial signaling channel and how meaning is applied to subsequent L2CAP channels using the AVDTP PSM. There can be at most signaling + 3 other channels according to AVDTP spec 1.3 section 5.4.6. Also see section 1.2 where it only talks about the signaling channel in singular rather than plural.
3. Even if it was theoretically possible to have multiple signaling channels for an ACL. I'm not aware of any devices that do that. Are you? Adding code complexity for non-existent theoretical situation is not a good practice IMO.
4. The array of contexts you have uses `CONFIG_BT_MAX_CONN`, implying that you really do want a 1:1 relationship. Otherwise it'd make more sense to define a new Kconfig option for this.

MarkWangChinese replied:
OK, @jhedberg Thanks, I see the follow line in the spec too. I will update the codes.
> The signaling channel (1) is only established once between two streaming devices.

MarkWangChinese replied:
updated. please help to review again.

jhedberg commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -30,50 +30,69 @@ LOG_MODULE_REGISTER(bt_avdtp);
-K_MUTEX_DEFINE(avdtp_mutex);
-#define AVDTP_LOCK() k_mutex_lock(&avdtp_mutex, K_FOREVER)
-#define AVDTP_UNLOCK() k_mutex_unlock(&avdtp_mutex)
+#define AVDTP_LOCK()   k_sem_take(&session->sem_lock, K_FOREVER)
+#define AVDTP_UNLOCK() k_sem_give(&session->sem_lock)
```
Please never create macros that make this kind of assumptions of the environment that they're invoked in, i.e. in this case that there has to be a given variable with a specific name. You should instead pass the session as a macro parameter. That said, I'm not sure what's the benefit of using macros here, since I don't think you want to preserve the possibility of redefining them or similar. Can't these just be one-line functions (which the compiler will certainly inline)?
```c
static void avdtp_lock(struct bt_avdtp *session)
{
	k_sem_take(&session->sem_lock);
}
```

MarkWangChinese replied:
Updated as you suggested.

jhedberg commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -30,50 +30,78 @@ LOG_MODULE_REGISTER(bt_avdtp);
-	AVDTP_STREAMING,
-	AVDTP_CLOSING,
-	AVDTP_ABORTING,
+	AVDTP_OPENING = 0x04U,
+	AVDTP_OPEN = 0x08U,
+	AVDTP_STREAMING = 0x10U,
+	AVDTP_CLOSING = 0x20U,
+	AVDTP_ABORTING = 0x40U,
```
These look like bits to me, why not use the `BIT()` macro? Actually, why are they bits? These seem internal (not something that goes over the air) and AFAIK you can't have two states at the same time.

MarkWangChinese replied:
yes, they are bits I will change as `BIT()`, and it can't be two states. I use it for easy `&` operation to check whether the current state is right state for the requested operation. For example: When closing the stream, the current state `AVDTP_OPEN `or `AVDTP_STREAMING` are both OK, I check it by `(param->sep->state & check_state)`.

MarkWangChinese replied:
Updated.

jhedberg commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -970,24 +1100,24 @@ int bt_avdtp_l2cap_recv(struct bt_l2cap_chan *chan, struct net_buf *buf)
-		.encrypt_change = bt_avdtp_l2cap_encrypt_changed,
-		.recv = bt_avdtp_l2cap_recv
-	};
+	static const struct bt_l2cap_chan_ops ops = {.connected = bt_avdtp_l2cap_connected,
+						     .disconnected = bt_avdtp_l2cap_disconnected,
+						     .encrypt_change =
+							     bt_avdtp_l2cap_encrypt_changed,
+						     .recv = bt_avdtp_l2cap_recv};
```
I think this is cleaner (hopefully clang-format will not fight against it):
```c
	static const struct bt_l2cap_chan_ops ops = {
		.connected = bt_avdtp_l2cap_connected,
		...
	};
```

MarkWangChinese replied:
Unfortunately clang-format fight against it. I think clang-format do this way because it doesn't exceed the max line length. there are your style of codes in avdtp.c too, but clang-format doesn't change them.

jhedberg replied:
Is it really failing CI, or you're just getting a change suggestion from clang-format?

MarkWangChinese replied:
I use the clang-format to format it locally, it changes as my codes. I am not sure whether the CI will fail, let me have a try later when I push new commit.

MarkWangChinese replied:
Because there are two places to use the same `ops`, so I move it out of function scope to file scope. and I use your style, CI doesn't report errors.

jhedberg commented:
@lylezhu2012 are you planning to review this?

lylezhu2012 commented:
> @lylezhu2012 are you planning to review this?

Yes. I will make it done before this weekend.

lylezhu2012 commented on subsys/bluetooth/host/classic/a2dp.c:
```diff
@@ -216,235 +178,212 @@ static int a2dp_set_config_ind(struct bt_avdtp *session, struct bt_avdtp_sep *se
+	req_cb = a2dp_cb != NULL ? a2dp_cb->release_req : NULL;
+	done_cb =
+		(ep->stream != NULL && ep->stream->ops != NULL) ? ep->stream->ops->released : NULL;
+	return a2dp_ctrl_ind(session, sep, errcode, req_cb, done_cb, false);
```
Whether the last parameter `clear_stream` should be `true`?

MarkWangChinese replied:
Updated.

lylezhu2012 commented on subsys/bluetooth/host/classic/a2dp.c:
```diff
@@ -216,235 +178,212 @@ static int a2dp_set_config_ind(struct bt_avdtp *session, struct bt_avdtp_sep *se
+	__ASSERT(sep, "Invalid sep");
+	req_cb = a2dp_cb != NULL ? a2dp_cb->abort_req : NULL;
+	done_cb = (ep->stream != NULL && ep->stream->ops != NULL) ? ep->stream->ops->aborted : NULL;
+	return a2dp_ctrl_ind(session, sep, errcode, req_cb, done_cb, false);
```
Same here. Whether the last parameter `clear_stream` should be `true`?

MarkWangChinese replied:
Updated.

lylezhu2012 commented on subsys/bluetooth/host/classic/a2dp.c:
```diff
@@ -703,40 +633,95 @@ int bt_a2dp_stream_config(struct bt_a2dp *a2dp, struct bt_a2dp_stream *stream,
+typedef void (*bt_a2dp_done_cb)(struct bt_a2dp_stream *stream);
+
+static int bt_a2dp_ctrl_cb(struct bt_avdtp_req *req, bt_a2dp_rsp_cb rsp_cb, bt_a2dp_done_cb done_cb,
+			   bool cler_stream)
```
Is it `clear_stream` instead of `cler_stream`?

MarkWangChinese replied:
Updated.

lylezhu2012 commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -343,26 +385,42 @@ static void avdtp_process_configuration(struct bt_avdtp *session,
 		uint8_t error_code = 0;
 
 		sep = avdtp_get_sep(net_buf_pull_u8(buf) >> 2);
```
To prevent malicious requests, it is best to determine the remaining data length before pulling the buffer.

MarkWangChinese replied:
Updated.

MarkWangChinese commented on subsys/bluetooth/host/classic/a2dp.c:
```diff
@@ -65,63 +63,26 @@ struct bt_a2dp {
 	uint8_t get_cap_index;
 	enum bt_a2dp_internal_state a2dp_state;
 	uint8_t peer_seps_count;
+	uint8_t used;
```
Deleted.

MarkWangChinese commented on subsys/bluetooth/host/classic/a2dp.c:
```diff
@@ -216,235 +178,212 @@ static int a2dp_set_config_ind(struct bt_avdtp *session, struct bt_avdtp_sep *se
+			*errcode = BT_AVDTP_BAD_ACP_SEID;
+			return -EINVAL;
+		}
+	} else if (!reconfig && (a2dp_cb == NULL || a2dp_cb->config_req == NULL)) {
```
Use `goto` to optimize it.

MarkWangChinese commented on subsys/bluetooth/host/classic/a2dp.c:
```diff
@@ -216,235 +178,212 @@ static int a2dp_set_config_ind(struct bt_avdtp *session, struct bt_avdtp_sep *se
+		return -EINVAL;
+	}
+
+	if ((a2dp_cb != NULL) && (!reconfig || (reconfig && a2dp_cb->reconfig_req != NULL))) {
```
Use `goto` to optimize it.

jhedberg replied:
You should probably check for `a2dp_cb != NULL` early in the function and return if it's NULL. That would simplify the if-conditions later in the function. Actually, does it make sense to let `a2dp_cb` be optional? Isn't it a bug in the application if it's not set?

jhedberg commented:
@MarkWangChinese there still seems to be some compliance error

jhedberg commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -1105,11 +1104,10 @@ int bt_avdtp_l2cap_recv(struct bt_l2cap_chan *chan, struct net_buf *buf)
+static const struct bt_l2cap_chan_ops signal_chan_ops = {.connected = bt_avdtp_l2cap_connected,
+							 .disconnected =
+								 bt_avdtp_l2cap_disconnected,
+							 .recv = bt_avdtp_l2cap_recv};
```
This doesn't help readability at all. Please revert this change (it will not cause CI failures, just warnings)

Thalley replied:
A good solution for this is to always end struct initialization with a final `,`.

Firstly that makes clang-format format it much better
Secondly, this makes adding additional lines to the struct initialization a single line instead of `,\n<new_line>\n`

jhedberg reviewed (changes requested):
A few comments. Generally, please try to keep style changes in separate commits. Right now you have many style changes mixed with functional changes in the first large commit, which makes it very hard to review it (to see what actually changed). Also, please don't blindly follow clang-format, since it often results in worse code. CI will not fail as long as you follow the Zephyr coding style - you will at most get clang-format warnings, but those can be ignored.

jhedberg commented:
Github doesn't allow doing inline commenting on commit messages, so I'll do it here. Let's look at the first commit in this PR:

>Implement the bt_a2dp_stream_reconfig: Modify reconfig_req callback to
pass codec_cfg to application. Remove reconfig_rsp callback,
config_rsp is used. Remove reconfigured callback,
configured callback is used.
> 
> Implement the bt_a2dp_stream_release: The release_req,
release_rsp and released callbacks are already defined,
implement them in this change.
> 
> Implement the bt_a2dp_stream_suspend: The suspend_req,
suspend_rsp and suspended callbacks are already defined,
implement them in this change.
> 
> Add and implement the bt_a2dp_stream_abort: Add abort_req,
abort_rsp and aborted callbacks.
> 
> Implement bt_a2dp_disconnect.
> 
> Simplify the codes: (1) use the same bt_avdtp_ctrl_params to process open,
start, suspend, close and abort operations. (2) use same a2dp_ctrl_ind to
process open, start, suspend, close and abort operations. (3) use the same
bt_a2dp_ctrl_cb to process open, start, suspend, close and abort
operations. (4) use the same bt_avdtp_ctrl to process open, start,
suspend, close and abort operations. (5) move the status to common struct
bt_avdtp_req.
> 
> change A2DP_MAX_IE_LENGTH as BT_A2DP_MAX_IE_LENGTH

Each of those paragraph is a single logical change in itself, i.e. these changes should all be in separate commits. Whenever you find yourself enumerating multiple changes in the commit message and those changes are not tightly coupled, it's a good indication that you need to split things up. Additionally, in this same commit you're also including coding style changes to unrelated lines of code - those should also be in a separate commit.

MarkWangChinese committed:
Bluetooth: A2DP: clang-format the codes

run clang-format

Signed-off-by: Mark Wang <yichang.wang@nxp.com>

MarkWangChinese committed:
Bluetooth: A2DP: implement reconfig and optimize codes

Implement the bt_a2dp_stream_reconfig: Modify reconfig_req callback to
pass codec_cfg to application. Remove reconfig_rsp callback,
config_rsp is used. Remove reconfigured callback,
configured callback is used.

move the status to common struct bt_avdtp_req,
use same bt_avdtp_ctrl_params to process control-like avdtp
cmds (start, open etc), use the same a2dp_ctrl_ind to process
control-like cmds (start, open etc), use the same
bt_a2dp_ctrl_cb to process control-like cmds (start, open etc),
use the same bt_avdtp_ctrl to process control-like cmds
(start, open etc), optimize getting a2dp conn by index,
use sem to replace mutex and optimze the lock codes to be
based on context/instance.

Signed-off-by: Mark Wang <yichang.wang@nxp.com>

MarkWangChinese committed:
Bluetooth: A2DP: implement close, suspend and abort

implement avdtp close, suspend and abort and the a2dp
interfaces.

Signed-off-by: Mark Wang <yichang.wang@nxp.com>

MarkWangChinese committed:
tests: bluetooth: shell: add a2dp new features test

add test for reconfigure, release, suspend, abort and disconnect.
app_config_req and app_reconfig_req always accept the req,
so don't need to handle reject case.

Signed-off-by: Mark Wang <yichang.wang@nxp.com>

MarkWangChinese commented:
> Github doesn't allow doing inline commenting on commit messages, so I'll do it here. Let's look at the first commit in this PR:
> 
> > Implement the bt_a2dp_stream_reconfig: Modify reconfig_req callback to
> > pass codec_cfg to application. Remove reconfig_rsp callback,
> > config_rsp is used. Remove reconfigured callback,
> > configured callback is used.
> > Implement the bt_a2dp_stream_release: The release_req,
> > release_rsp and released callbacks are already defined,
> > implement them in this change.
> > Implement the bt_a2dp_stream_suspend: The suspend_req,
> > suspend_rsp and suspended callbacks are already defined,
> > implement them in this change.
> > Add and implement the bt_a2dp_stream_abort: Add abort_req,
> > abort_rsp and aborted callbacks.
> > Implement bt_a2dp_disconnect.
> > Simplify the codes: (1) use the same bt_avdtp_ctrl_params to process open,
> > start, suspend, close and abort operations. (2) use same a2dp_ctrl_ind to
> > process open, start, suspend, close and abort operations. (3) use the same
> > bt_a2dp_ctrl_cb to process open, start, suspend, close and abort
> > operations. (4) use the same bt_avdtp_ctrl to process open, start,
> > suspend, close and abort operations. (5) move the status to common struct
> > bt_avdtp_req.
> > change A2DP_MAX_IE_LENGTH as BT_A2DP_MAX_IE_LENGTH
> 
> Each of those paragraphs is a single logical change in itself, i.e. these changes should all be in separate commits. Whenever you find yourself enumerating multiple changes in the commit message and those changes are not tightly coupled, it's a good indication that you need to split things up. Additionally, in this same commit you're also including coding style changes to unrelated lines of code - those should also be in a separate commit.

Hi @jhedberg I have split the commit to three commits. I do the code format firstly with one commit. Then add one commit for the reconfig implementation and codes optimization. The last one is the new added close, suspend and abort. The second commit is still little big, but it is hard and need more effort to split because there are some common optimizations in it. Please help to review again. Thanks.

jhedberg commented:
@MarkWangChinese are you still working on this? There are several failed CI checks.

MarkWangChinese commented:
> @MarkWangChinese are you still working on this? There are several failed CI checks.

Hi @jhedberg Sure, I am still working on it. I will fix it recently. Thanks

jhedberg commented on subsys/bluetooth/host/classic/shell/a2dp.c:
```diff
@@ -324,6 +324,25 @@ int app_config_req(struct bt_a2dp *a2dp, struct bt_a2dp_ep *ep,
+int app_reconfig_req(struct bt_a2dp_stream *stream,
+	struct bt_a2dp_codec_cfg *codec_cfg, uint8_t *rsp_err_code)
+{
+	*rsp_err_code = 0;
+
+	shell_print(ctx_shell, "receive requesting reconfig and accept");
+	if (*rsp_err_code == 0) {
```
Is there some call missing here that should set `*rsp_err_code` to something else than 0? Otherwise the if statement is redundant since you're setting it to 0 right before.

MarkWangChinese replied:
Updated. This callback will always accept the `req`, so it set `*rsp_err_code` as 0. remove the redundant codes.

lylezhu2012 commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -299,61 +363,77 @@ static void avdtp_get_capabilities_handler(struct bt_avdtp *session, struct net_
+				expected_state = AVDTP_IDLE;
+			}
+
+			avdtp_sep_lock(sep);
```
Hierarchically, this `lock` has no corresponding `unlock`.

MarkWangChinese replied:
Fixed

MarkWangChinese commented on subsys/bluetooth/host/classic/avdtp.c:
```diff
@@ -945,21 +1118,20 @@ int bt_avdtp_l2cap_recv(struct bt_l2cap_chan *chan, struct net_buf *buf)
 		LOG_DBG("Error: Session not valid");
-		return 0;
+		return -EINVAL;
 	}
```
Fixed

jhedberg commented on subsys/bluetooth/host/classic/shell/a2dp.c:
```diff
@@ -167,7 +167,7 @@ static struct bt_sdp_attribute a2dp_source_attrs[] = {
 			{
 				BT_SDP_TYPE_SIZE(BT_SDP_UINT16),
-				BT_SDP_ARRAY_16(0X0100u)
+				BT_SDP_ARRAY_16(0X0100U)
```
Why not fix the `X` as well?

MarkWangChinese replied:
Fixed

jhedberg commented on subsys/bluetooth/host/classic/shell/a2dp.c:
```diff
@@ -98,7 +98,7 @@ static struct bt_sdp_attribute a2dp_sink_attrs[] = {
 			{
 				BT_SDP_TYPE_SIZE(BT_SDP_UINT16), /* 09 */
-				BT_SDP_ARRAY_16(0X0100u) /* AVDTP version: 01 00 */
+				BT_SDP_ARRAY_16(0X0100U) /* AVDTP version: 01 00 */
```
Same here.

MarkWangChinese replied:
Fixed

jhedberg reviewed (commented):
This looks generally good enough to be merged. Just a few comments regarding `x` vs `X` in hex values (not sure why you want `X` when almost everywhere the code uses `x`)
